### PR TITLE
Failure net crypto

### PR DESCRIPTION
**Pull request by NamsooCho**

### CODE REVIEW

File: src/toxcore/dht/server/errors.rs [hunk 1]
```diff
@@ -71,6 +72,9 @@ pub enum HandlePacketErrorKind {
     /// Error indicates that NetCrypto is not initialized.
     #[fail(display = "NetCrypto is not initialized error")]
     NetCrypto,
+    /// Error indicates that handling NetCrypto packet made an error.
+    #[fail(display = "Handling NetCrypto packet has error")]
```
Comment (kpp) [inline src/toxcore/dht/server/errors.rs]: Handling NetCrypto packet failed

File: src/toxcore/net_crypto/errors.rs [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+    InvalidRealPk,
+    /// Error indicates that cookie contains invalid dht pk
+    #[fail(display = "Cookie contains invalid dht pk")]
+    InvalidDhtPk,
+    /// Error indicates that there is PacketsArrayError
+    #[fail(display = "PacketsArrayError occurs")]
+    PacketsArray,
+    /// Error indicates that real data is empty
+    #[fail(display = "Real data is empty")]
+    Empty,
```
Comment (kpp) [inline src/toxcore/net_crypto/errors.rs]: DataEmpty

File: src/toxcore/net_crypto/errors.rs [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+    CookieTimedOut,
+    /// Error indicates that cookie contains invalid real pk
+    #[fail(display = "Cookie contains invalid real pk")]
+    InvalidRealPk,
+    /// Error indicates that cookie contains invalid dht pk
+    #[fail(display = "Cookie contains invalid dht pk")]
+    InvalidDhtPk,
+    /// Error indicates that there is PacketsArrayError
+    #[fail(display = "PacketsArrayError occurs")]
+    PacketsArray,
```
Comment (kpp) [inline src/toxcore/net_crypto/errors.rs]: PacketsArrayError

File: src/toxcore/net_crypto/errors.rs [hunk 4]
```diff
@@ -0,0 +1,612 @@
...
+    },
+    /// Error indicates that no crypto connection for address.
+    #[fail(display = "No crypto connection for address: {:?}", addr)]
+    NoConnection {
+        /// The address for connection which don't exist
+        addr: SocketAddr,
+    },
+    /// Error indicates that invalid SHA512 hash of cookie
+    #[fail(display = "Invalid SHA512 hash of cookie")]
+    Sha512,
```
Comment (kpp) [inline src/toxcore/net_crypto/errors.rs]: BadSha512

File: src/toxcore/net_crypto/errors.rs [hunk 5]
```diff
@@ -0,0 +1,612 @@
...
+    GetPayload,
+    /// Error indicates that sending response packet error.
+    #[fail(display = "Sending response error")]
+    SendTo,
+    /// Error indicates that sending lossless response packet error.
+    #[fail(display = "Sending lossless response error")]
+    SendToLossless,
+    /// Error indicates that NetCrypto can't handle packet in current connection state.
+    #[fail(display = "Can't handle CookieResponse in current connection state")]
+    CannotHandle,
```
Comment (kpp) [inline src/toxcore/net_crypto/errors.rs]: CannotHandle - what's that for?

Comment (NamsooCho) [reply src/toxcore/net_crypto/errors.rs]: Changed it to `InvalidState`

File: src/toxcore/net_crypto/errors.rs [hunk 6]
```diff
@@ -0,0 +1,612 @@
...
+    use crate::toxcore::io_tokio::*;
+
+    use std::time::Duration;
+    use futures::future::Future;
+
+    #[test]
+    fn packets_array_error() {
+        let error = PacketsArrayError::from(PacketsArrayErrorKind::ArrayFull);
+        assert!(error.cause().is_none());
+        assert!(error.backtrace().is_some());
```
Comment (kurnevsky) [inline src/toxcore/net_crypto/errors.rs]: Isn't bactrace optional?

File: src/toxcore/net_crypto/mod.rs [hunk 7]
```diff
@@ -253,75 +249,60 @@ impl NetCrypto {
...
                 send_to(&self.dht_pk_tx, (connection.peer_real_pk, cookie.dht_pk))
-                    .map_err(|e|
-                        Error::new(ErrorKind::Other,
-                            format!("Failed to send packet: {:?}", e)
-                    ))
-                    .and_then(|()| future::err(Error::new(
-                        ErrorKind::Other,
-                        "Cookie contains invalid dht pk"
-                    )))
+                    .map_err(|_|
```
Comment (kurnevsky) [inline src/toxcore/net_crypto/mod.rs]: Why this error is ignored?

File: src/toxcore/net_crypto/errors.rs (hunk 6 above)
Comment (NamsooCho) [reply src/toxcore/net_crypto/errors.rs]: Yes. But we introduced `backtrace` to find the location of error.
I think we should enforce to use `backtrace`.

Comment (kpp) [reply src/toxcore/net_crypto/errors.rs]: No, backtrace is optional

File: src/toxcore/net_crypto/mod.rs (hunk 7 above)
Comment (NamsooCho) [reply src/toxcore/net_crypto/mod.rs]: done

**Pull request merged**